### DISCUSSION TITLE
Unable to backup kafka data from persistent volumes

### DISCUSSION THREAD

**Discussion by shreyasarani23** [strimzi/strimzi-kafka-operator#4892] [General]
I have created one storage class [kafka-storage-class-> name of the storage class]. And the storage class name is specified in kafka deployment file. I have given three replicas for kafka and zookeeper so it is creating 6 persistent volumes and 6 pvc for kafka and zookeeper pods. Basically creating 6 azure disks. When i delete the kafka cluster and redploy it, the data in the topics is lost. But the persistent volumes remain as it is, because the reclaim policy is given as true for storage class. But I didn't know how to restore kafka data from PV's. I have referred this link https://strimzi.io/docs/operators/latest/full/using.html#cluster-recovery_str. But couldn't restore the data.
Can you please specify the steps in detail to restore kafka data from persistent voulmes? I have attached my kafka deployment file please help me to restore kafka data from persistent volumes
[kafka.yaml.txt](https://github.com/strimzi/strimzi-kafka-operator/files/6420615/kafka.yaml.txt)

**Comment by scholzj** [comment]
What version of Strimzi do you use?

**Reply by shreyasarani23** [reply]
latest version of strimzi [strimzi-0.22.1]

**Reply by scholzj** [reply]
So what steps have you done from the guide and where did it stopped working for you?

**Reply by shreyasarani23** [reply]
In step 3 it is says the following
Recreate the original PVC resource specifications, linking the PVCs to the appropriate PV:

For example:

apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: data-0-my-cluster-kafka-0
spec:
  accessModes:
  - ReadWriteOnce
  resources:
    requests:
      storage: 100Gi
  storageClassName: gp2-retain
  volumeMode: Filesystem
  volumeName: pvc-7e1f67f9-3317-11ea-a650-06

But I haven't created the PV and PVC's manually strimzi automatically created it for me. So how do I edit the Persistent volume claim?
 And also it says the following
The recovery procedure is the same as with namespace deletion, assuming PVs can be recovered and they were created manually.

I haven't created PV's and PVC's manually strimzi automatically generated it for me.

**Reply by scholzj** [reply]
Well, they were created by strimzi originally, but that was when you properly created the cluster. Now you are trying to do a recovery after some kind of disaster. So you have to do it manually. Strimzi does not know which PVs or which infrastructure volumes are what, so you need to decide that and create everything up to PVCs. Once the PVCs exist with the right name, Strimzi can just use them.

**Reply by shreyasarani23** [reply]
1. So now I should create 6 pvc's  manually , linking the PVCs to the appropriate PV right?
2. And in step 4 it says the following 
Edit the PV specifications to delete the claimRef properties that bound the original PVC.
example:
```
apiVersion: v1
kind: PersistentVolume
metadata:
  annotations:
    kubernetes.io/createdby: aws-ebs-dynamic-provisioner
    pv.kubernetes.io/bound-by-controller: "yes"
    pv.kubernetes.io/provisioned-by: kubernetes.io/aws-ebs
  creationTimestamp: "<date>"
  finalizers:
  - kubernetes.io/pv-protection
  labels:
    failure-domain.beta.kubernetes.io/region: eu-west-1
    failure-domain.beta.kubernetes.io/zone: eu-west-1c
  name: pvc-7e226978-3317-11ea-97b0-0aef8816c7ea
  resourceVersion: "39431"
  selfLink: /api/v1/persistentvolumes/pvc-7e226978-3317-11ea-97b0-0aef8816c7ea
  uid: 7efe6b0d-3317-11ea-a650-06e1eadd9a4c
spec:
  accessModes:
  - ReadWriteOnce
  awsElasticBlockStore:
    fsType: xfs
    volumeID: aws://eu-west-1c/vol-09db3141656d1c258
  capacity:
    storage: 100Gi
  claimRef:
    apiVersion: v1
    kind: PersistentVolumeClaim
    name: data-0-my-cluster-kafka-2
    namespace: myproject
    resourceVersion: "39113"
    uid: 54be1c60-3319-11ea-97b0-0aef8816c7ea
  nodeAffinity:
    required:
      nodeSelectorTerms:
      - matchExpressions:
        - key: failure-domain.beta.kubernetes.io/zone
          operator: In
          values:
          - eu-west-1c
        - key: failure-domain.beta.kubernetes.io/region
          operator: In
          values:
          - eu-west-1
  persistentVolumeReclaimPolicy: Retain
  storageClassName: gp2-retain
  volumeMode: Filesystem
```

So now for all my  6 Persistent volume I should remove the claimRef section right?

**Reply by scholzj** [reply]
Yes, that is correct. That identifies the claim it was bound to and it will not bind to the new claim while it is still there IIRC.

**Reply by shreyasarani23** [reply]
But how to I delete the claimRef section of the persistent volume?
Below is the command I used to get the yaml config of my persistent volume.
shreyas@Azure:~/kafka-deployment$ kubectl get pv/pvc-d4779493-f312-459f-86f1-f31239a570e3 -o yaml
 
So it shows me the yaml config of persistent volume.

Now how do I delete it ? should I copy that yaml config to a file and remove the claim ref section and deploy it using the command
kubectl apply -f pv.yaml?

or is there any other method to delete the ClaimRef section?

**Reply by scholzj** [reply]
I think when I was testing it I just did `kubectl edit pv ...` and removed it. Unless you need to script it, that might be the easiest way.

**Reply by shreyasarani23** [reply]
Another doubt is that after manually creating the PVC's we should first deploy the pvc's or first delete the claimref section in persistent volume and then deploy the pvc's ? please can you clarify?
Is this command correct for editing the PV?
 kubectl edit pv/pv-name -o yaml --save-config
And can you give me the full command to edit the PV?

**Reply by shreyasarani23** [reply]
And I didn't understand step 6 it says the following
Recreate your cluster.

Follow the steps depending on whether or not you have all the KafkaTopic resources needed to recreate your cluster.

Option 1: If you have all the KafkaTopic resources that existed before you lost your cluster, including internal topics such as committed offsets from __consumer_offsets:

Recreate all KafkaTopic resources.

It is essential that you recreate the resources before deploying the cluster, or the Topic Operator will delete the topics.

Deploy the Kafka cluster.

For example:

kubectl apply -f kafka.yaml

what do u mean by kafkatopic resources and how to check if the kafkatopic resources exists or not?
How do u recreate all the KafkaTopic resources?

And can you explain when should we go to option2 i.e. If you do not have all the KafkaTopic resources that existed before you lost your cluster

**Reply by scholzj** [reply]
> Another doubt is that after manually creating the PVC's we should first deploy the pvc's or first delete the claimref section in persistent volume and then deploy the pvc's ? please can you clarify?

The documentation says first create the PVC and then remove the reference. 

> Is this command correct for editing the PV?
kubectl edit pv/pv-name -o yaml --save-config
And can you give me the full command to edit the PV?

I normally edit them with something like this: `kubectl edit pv pv-name`

> what do u mean by kafkatopic resources and how to check if the kafkatopic resources exists or not?
How do u recreate all the KafkaTopic resources?

`KafkaTopic` resources are what the topics operator uses for creating and managing topics. So you either created them and have the YAMLs or Strimzi created them in which case you can do `kubectl get KafkaTopic`. If you do not know what they are, you might want to follow the second option.

**Reply by shreyasarani23** [reply]
Hi I am following the Option 2: If you do not have all the KafkaTopic resources that existed before you lost your cluster:
1. As a first step I deployed the kafka cluster without the topic operator.
2. But I am stuck in the second step. I am getting the following error.
shreyas@Azure:~/kafka-deployment$ kubectl run kafka-admin -ti --image=quay.io/strimzi/kafka:0.22.1-kafka-2.7.0  --rm=true --restart=Never -- ./bin/kafka-topics.sh --bootstrap-server my-kafka-cluster-kafka-bootstrap:9092 --topic __strimzi-topic-operator-kstreams-topic-store-changelog --delete && ./bin/kafka-topics.sh --bootstrap-server my-kafka-cluster-kafka-bootstrap:9092 --topic __strimzi_store_topic --delete
If you don't see a command prompt, try pressing enter.
[2021-05-05 07:06:53,197] WARN Couldn't resolve server my-kafka-cluster-kafka-bootstrap:9092 from bootstrap.servers as DNS resolution failed for my-kafka-cluster-kafka-bootstrap (org.apache.kafka.clients.ClientUtils)
Exception in thread "main" org.apache.kafka.common.KafkaException: Failed to create new KafkaAdminClient
        at org.apache.kafka.clients.admin.KafkaAdminClient.createInternal(KafkaAdminClient.java:535)
        at org.apache.kafka.clients.admin.Admin.create(Admin.java:65)
        at kafka.admin.TopicCommand$AdminClientTopicService$.createAdminClient(TopicCommand.scala:228)
        at kafka.admin.TopicCommand$AdminClientTopicService$.apply(TopicCommand.scala:232)
        at kafka.admin.TopicCommand$.main(TopicCommand.scala:59)
        at kafka.admin.TopicCommand.main(TopicCommand.scala)
Caused by: org.apache.kafka.common.config.ConfigException: No resolvable bootstrap urls given in bootstrap.servers
        at org.apache.kafka.clients.ClientUtils.parseAndValidateAddresses(ClientUtils.java:89)
        at org.apache.kafka.clients.ClientUtils.parseAndValidateAddresses(ClientUtils.java:48)
        at org.apache.kafka.clients.admin.KafkaAdminClient.createInternal(KafkaAdminClient.java:489)
        ... 5 more
pod "kafka-admin" deleted
pod default/kafka-admin terminated (Error)

Please Can you help me to resolve this error?

**Reply by scholzj** [reply]
It says _No resolvable bootstrap urls given in bootstrap.servers_. So it cannot resolve the address of the bootstrap server you gave it. Was the Kafka cluster deployed? Is it running? If yes, double check that the address is correct and if it is you will need to check your DNS etc.

**Reply by shreyasarani23** [reply]
Yes the kafka cluster is deployed and it is running and the address is correct. Below is the command to check bootstrap service
shreyas@Azure:~$ kubectl get svc -n kafka
NAME                                TYPE        CLUSTER-IP     EXTERNAL-IP   PORT(S)                      AGE
my-kafka-cluster-kafka-bootstrap    ClusterIP   10.0.96.252    <none>        9091/TCP,9092/TCP            86m
my-kafka-cluster-kafka-brokers      ClusterIP   None           <none>        9091/TCP,9092/TCP            86m
my-kafka-cluster-zookeeper-client   ClusterIP   10.0.248.202   <none>        2181/TCP                     87m
my-kafka-cluster-zookeeper-nodes    ClusterIP   None           <none>        2181/TCP,2888/TCP,3888/TCP   87m
shreyas@Azure:~$ kubectl run kafka-admin -ti --image=quay.io/strimzi/kafka:0.22.1-kafka-2.7.0  --rm=true --restart=Never -- ./bin/kafka-topics.sh --bootstrap-server my-kafka-cluster-kafka-bootstrap:9092 --topic __strimzi-topic-operator-kstreams-topic-store-changelog --delete && ./bin/kafka-topics.sh --bootstrap-server my-kafka-cluster-kafka-bootstrap:9092 --topic __strimzi_store_topic --delete
If you don't see a command prompt, try pressing enter.
[2021-05-05 08:32:16,383] WARN Couldn't resolve server my-kafka-cluster-kafka-bootstrap:9092 from bootstrap.servers as DNS resolution failed for my-kafka-cluster-kafka-bootstrap (org.apache.kafka.clients.ClientUtils)
Exception in thread "main" org.apache.kafka.common.KafkaException: Failed to create new KafkaAdminClient
        at org.apache.kafka.clients.admin.KafkaAdminClient.createInternal(KafkaAdminClient.java:535)
        at org.apache.kafka.clients.admin.Admin.create(Admin.java:65)
        at kafka.admin.TopicCommand$AdminClientTopicService$.createAdminClient(TopicCommand.scala:228)
        at kafka.admin.TopicCommand$AdminClientTopicService$.apply(TopicCommand.scala:232)
        at kafka.admin.TopicCommand$.main(TopicCommand.scala:59)
        at kafka.admin.TopicCommand.main(TopicCommand.scala)
Caused by: org.apache.kafka.common.config.ConfigException: No resolvable bootstrap urls given in bootstrap.servers
        at org.apache.kafka.clients.ClientUtils.parseAndValidateAddresses(ClientUtils.java:89)
        at org.apache.kafka.clients.ClientUtils.parseAndValidateAddresses(ClientUtils.java:48)
        at org.apache.kafka.clients.admin.KafkaAdminClient.createInternal(KafkaAdminClient.java:489)
        ... 5 more
pod "kafka-admin" deleted
pod default/kafka-admin terminated (Error)

Still getting the same error.
How do I check my dns??

**Reply by scholzj** [reply]
Aren't you running in in a different namespace?

**Reply by shreyasarani23** [reply]
yes I am running my kafka cluster in separate namespace called kafka

**Reply by scholzj** [reply]
But using `<service-name>:<port>` works only in the same namespace. If you connect from other namespace you need to use something like `<service-name>.<namespace>.svc:<port>`.

**Reply by shreyasarani23** [reply]
ohk let me try with the above

**Reply by shreyasarani23** [reply]
shreyas@Azure:~/kafka-deployment$ kubectl run kafka-admin -ti --image=quay.io/strimzi/kafka:0.22.1-kafka-2.7.0  --rm=true --restart=Never -- ./bin/kafka-topics.sh --bootstrap-server my-kafka-cluster-kafka-bootstrap.kafka.svc:9092 --topic __strimzi-topic-operator-kstreams-topic-store-changelog --delete && ./bin/kafka-topics.sh --bootstrap-server my-kafka-cluster-kafka-bootstrap.kafka.svc:9092 --topic __strimzi_store_topic --delete
If you don't see a command prompt, try pressing enter.
pod "kafka-admin" deleted
bash: ./bin/kafka-topics.sh: No such file or directory

getting this error I don't why.

**Reply by scholzj** [reply]
I don't know why either. It looks fine to me.

**Reply by shreyasarani23** [reply]
Any work around for this??

**Reply by shreyasarani23** [reply]
tried once more now getting this error
shreyas@Azure:~/kafka-deployment$ kubectl run kafka-admin -ti --image=quay.io/strimzi/kafka:latest-kafka-2.7.0 --rm=true --restart=Never -- ./bin/kafka-topics.sh --bootstrap-server my-kafka-cluster-kafka-bootstrap.kafka.svc:9092 --topic __strimzi-topic-operator-kstreams-topic-store-changelog --delete && ./bin/kafka-topics.sh --bootstrap-server my-kafka-cluster-kafka-bootstrap.kafka.svc:9092 --topic __strimzi_store_topic --delete
If you don't see a command prompt, try pressing enter.
Error while executing topic command : Topic '__strimzi-topic-operator-kstreams-topic-store-changelog' does not exist as expected
[2021-05-05 10:14:13,889] ERROR java.lang.IllegalArgumentException: Topic '__strimzi-topic-operator-kstreams-topic-store-changelog' does not exist as expected
        at kafka.admin.TopicCommand$.kafka$admin$TopicCommand$$ensureTopicExists(TopicCommand.scala:539)
        at kafka.admin.TopicCommand$AdminClientTopicService.deleteTopic(TopicCommand.scala:357)
        at kafka.admin.TopicCommand$.main(TopicCommand.scala:72)
        at kafka.admin.TopicCommand.main(TopicCommand.scala)
 (kafka.admin.TopicCommand$)
pod "kafka-admin" deleted
pod default/kafka-admin terminated (Error)

**Reply by shreyasarani23** [reply]
And One more thing I didn't create topic through yaml file in which you specify kind:KafkaTopic. In my kafka deployment  I have set 
auto.create.topics.enable=true. And directly sending messages using the below command
kubectl -n kafka run kafka-producer -ti --image=quay.io/strimzi/kafka:0.22.1-kafka-2.7.0 --rm=true --restart=Never -- bin/kafka-console-producer.sh --broker-list my-kafka-cluster-kafka-bootstrap:9092 --topic my-topic

**Reply by scholzj** [reply]
If the topic doesn't exist you probably already deleted it or you never had it in the first place.

**Reply by shreyasarani23** [reply]
No the topic exists . I went inside the pod and checked it .
1. Firstly I deployed kafka cluster without topic operator.
2. And  i went inside the pod and checked the topic exists or not .
shreyas@Azure:~/kafka-deployment$ kubectl exec -it my-kafka-cluster-kafka-0 -n kafka -- /bin/bash
[kafka@my-kafka-cluster-kafka-0 kafka]$ cd /
[kafka@my-kafka-cluster-kafka-0 /]$ cd var/lib/kafka/data/kafka-log0
[kafka@my-kafka-cluster-kafka-0 kafka-log0]$ ls
__consumer_offsets-0   __consumer_offsets-2   __consumer_offsets-30  __consumer_offsets-41  __consumer_offsets-8
__consumer_offsets-1   __consumer_offsets-20  __consumer_offsets-31  __consumer_offsets-42  __consumer_offsets-9
__consumer_offsets-10  __consumer_offsets-21  __consumer_offsets-32  __consumer_offsets-43  __strimzi-topic-operator-kstreams-topic-store-changelog-0
__consumer_offsets-11  __consumer_offsets-22  __consumer_offsets-33  __consumer_offsets-44  __strimzi_store_topic-0
__consumer_offsets-12  __consumer_offsets-23  __consumer_offsets-34  __consumer_offsets-45  cleaner-offset-checkpoint
__consumer_offsets-13  __consumer_offsets-24  __consumer_offsets-35  __consumer_offsets-46  log-start-offset-checkpoint
__consumer_offsets-14  __consumer_offsets-25  __consumer_offsets-36  __consumer_offsets-47  meta.properties
__consumer_offsets-15  __consumer_offsets-26  __consumer_offsets-37  __consumer_offsets-48  my-topic-0
__consumer_offsets-16  __consumer_offsets-27  __consumer_offsets-38  __consumer_offsets-49  recovery-point-offset-checkpoint
__consumer_offsets-17  __consumer_offsets-28  __consumer_offsets-39  __consumer_offsets-5   replication-offset-checkpoint
__consumer_offsets-18  __consumer_offsets-29  __consumer_offsets-4   __consumer_offsets-6
__consumer_offsets-19  __consumer_offsets-3   __consumer_offsets-40  __consumer_offsets-7

my-topic is the name of the topic and it has only one partition.
if u see the changelog directory the name of changelog directory is:
__strimzi-topic-operator-kstreams-topic-store-changelog-0

And in the command we are giving just __strimzi-topic-operator-kstreams-topic-store-changelog. So I guess that's the reason the above error is coming. And similarly for __strimzi_store_topic-0

**Reply by scholzj** [reply]
Then you should be able to delete them.

**Reply by shreyasarani23** [reply]
But if I delete __strimzi-topic-operator-kstreams-topic-store-changelog-0 and __strimzi_store_topic-0 and redeploy my cluster with topic operator it will delete the topic right?

**Reply by scholzj** [reply]
No, these topics are used by the topic operator - so it will recreate it but it will be empty.

**Reply by shreyasarani23** [reply]
But I want the data to be there and don't want to create the empty So what Should I do now?
My objective is to retrieve data from disaster and if I deploy topic operator now, the topics will be recreated and there won't be any data. I don't want this to happen.

**Reply by scholzj** [reply]
`__strimzi-topic-operator-kstreams-topic-store-changelog` and `__strimzi_store_topic` are strimzi internal topics. If you do not delete them you might loose the other topics.

**Reply by shreyasarani23** [reply]
Sorry I didn't understand the concept properly can u explain once again?
Right now there is data in the topic but the problem is when I deploy the kafka cluster with topic operator it will recreate the topic and there won't be any data. So I am asking is there a way to retain data even after deploying topic operator??

**Reply by scholzj** [reply]
Why do you need to retain the data of the internal topics belonging to the topic operator?

**Reply by shreyasarani23** [reply]
I am not speaking about the data of internal topics and I am talking about the topic which I have created. I have created a topic called my-topic. Now this my-topic contains some data but when I deploy the topic operator the data inside the my-topic will be deleted right?

**Reply by scholzj** [reply]
No. You have to delete the two internal topic first, then enable the topic operator and that way your `my-topic` will not be deleted. That is exactly what the documentation says.

**Reply by shreyasarani23** [reply]
But unable to delete the two internal topics
shreyas@Azure:~$ kubectl run kafka-admin -ti --image=quay.io/strimzi/kafka:latest-kafka-2.7.0 --rm=true --restart=Never -- ./bin/kafka-topics.sh --bootstrap-server my-kafka-cluster-kafka-bootstrap.kafka.svc:9092 --topic __strimzi-topic-operator-kstreams-topic-store-changelog-0 --delete && ./bin/kafka-topics.sh --bootstrap-server my-kafka-cluster-kafka-bootstrap.kafka.svc:9092 --topic __strimzi_store_topic-0 --delete
If you don't see a command prompt, try pressing enter.
Error while executing topic command : Topic '__strimzi-topic-operator-kstreams-topic-store-changelog-0' does not exist as expected
[2021-05-06 08:46:36,447] ERROR java.lang.IllegalArgumentException: Topic '__strimzi-topic-operator-kstreams-topic-store-changelog-0' does not exist as expected
        at kafka.admin.TopicCommand$.kafka$admin$TopicCommand$$ensureTopicExists(TopicCommand.scala:539)
        at kafka.admin.TopicCommand$AdminClientTopicService.deleteTopic(TopicCommand.scala:357)
        at kafka.admin.TopicCommand$.main(TopicCommand.scala:72)
        at kafka.admin.TopicCommand.main(TopicCommand.scala)
 (kafka.admin.TopicCommand$)
pod "kafka-admin" deleted
pod default/kafka-admin terminated (Error)

I don't know why I am facing this issue.

**Reply by shreyasarani23** [reply]
What happens if we deploy topic operator without deleting these two internal topics?

**Reply by shreyasarani23** [reply]
Hi @scholzj i was able to delete the internal topic by entering into the pod with the following command
[kafka@my-kafka-cluster-kafka-0 kafka]$ bin/kafka-topics.sh --bootstrap-server my-kafka-cluster-kafka-bootstrap:9092 --topic __strimzi_store_topic --delete
[kafka@my-kafka-cluster-kafka-0 kafka]$ bin/kafka-topics.sh --list --bootstrap-server my-kafka-cluster-kafka-bootstrap:9092
__consumer_offsets
my-topic

We can see that __Strimzi_store_topic is deleted. And I deployed my kafka cluster with topic operator. I was able to get the topic using the command kubectl get KafkaTopic. To check whether the data is present in the topic or not I executed the consumer pod and I was able to consume the records.

Thanks @scholzj for assisting me. Looking forward to work more on Strimzi.

**Reply by scholzj** [reply]
👍
